### PR TITLE
pytorch/cuda compatibility updates

### PR DESCRIPTION
**Pull request by dribnet**
Not necessarily recommending this is merged into the codebase - but thought I would offer some pytorch / CUDA compatibility changes taken from the upstream project in case they are useful for anyone else. In my case it allows me to run this project on an *older* version of pytorch (2.4.0) / cuda (12.1) [don't ask 😅] - but replacing the adam-atan2 dependency is also reported to be helpful running on *newer* versions of cuda as well (RTX 5090 / 4090 / 3090 with cuda 12.8).

Anyway, it worked for me - my first run of suduku even had a test accuracy of 93.5% (7% better than the paper?!)

```bash
wandb: Run summary:                                                                                     
wandb:            num_params 5028866                                                                    
wandb:        train/accuracy 0.98872                                                                    
wandb:           train/count 1                                                                          
wandb:  train/exact_accuracy 0.93514                                                                    
wandb:         train/lm_loss 0.50523                                                                    
wandb:              train/lr 0.0001                                                                     
wandb: train/q_halt_accuracy 0.97297                                                                    
wandb:     train/q_halt_loss 0.02499                                                                    
wandb:           train/steps 3.97297                                                                    
```

(If this does look useful but messy I can also clean it up a bit if you'd like to merge it in.)

------- 8< -------

replaced adam-atan2 with adam-atan2-pytorch which works on older and newer versions of CUDA / pytorch

https://github.com/sapientinc/HRM/issues/45

Also replaced nn.Buffer with register_buffer

https://github.com/sapientinc/HRM/pull/30

### CODE REVIEW

**Comment by olivkoch** [comment]
My repo has evaluation scripts and visualizations: https://github.com/olivkoch/nano-trm
